File: app/grading.py
"""Pure grading/validation logic. No tkinter dependency, easy to unit test."""

from app.constants import OPTIONS


def normalize_answer_string(raw_ans):
    return raw_ans.strip().upper().replace(" ", "")


def validate_answer_string(raw_ans, total=None):
    """Return None if raw_ans is a valid answer string, else an error message.

    When `total` is given, raw_ans must be exactly that many characters (used
    to check submitted answers against an already-registered key's length).
    When omitted, any non-empty string of valid option characters is
    accepted — the key's own length becomes the question count to grade.
    """
    if not raw_ans:
        return "정답을 입력해주세요."
    if total is not None and len(raw_ans) != total:
        return f"정답은 정확히 {total}자여야 합니다. (현재: {len(raw_ans)}자)"
    for char in raw_ans:
        if char not in OPTIONS:
            return f"유효하지 않은 문자: '{char}'"
    return None


def parse_answer_key(raw_ans):
    """Convert a validated answer string into a {question_no: answer} dict (1-indexed)."""
    return {i + 1: char for i, char in enumerate(raw_ans)}


def grade(user_answers, answer_key, total=None):
    """Return (correct_count, wrong_questions) comparing question numbers 1..total.

    `total` defaults to the number of questions in `answer_key`.
    """
    if total is None:
        total = len(answer_key)

    correct_count = 0
    wrong_questions = []
    for q in range(1, total + 1):
        if user_answers[q] == answer_key[q]:
            correct_count += 1
        else:
            wrong_questions.append(q)
    return correct_count, wrong_questions


def build_result_text(
    *, timestamp, correct_count, total, user_answers, answer_key, wrong_questions, notes
):
    """Build the plain-text report content written to the result .txt file."""
    accuracy = (correct_count / total) * 100
    lines = [
        "=========================================",
        "           토익 채점 및 오답 노트          ",
        "=========================================",
        f"일시: {timestamp}",
        f"점수: {correct_count} / {total} ({accuracy:.1f}%)",
        "",
        "[ 전체 답안 현황 ]",
    ]

    for i in range(1, total + 1):
        is_correct = "O" if user_answers[i] == answer_key[i] else "X"
        lines.append(f"{i:03d}번 | 제출: {user_answers[i]} | 정답: {answer_key[i]} | [{is_correct}]")

    lines.append("")
    lines.append("[ 오답 노트 ]")
    if wrong_questions:
        for q in wrong_questions:
            reason = notes.get(q, "").strip() or "메모 없음"
            lines.append(f"- {q:03d}번 (제출: {user_answers[q]} / 정답: {answer_key[q]})")
            lines.append(f"  이유: {reason}")
    else:
        lines.append("틀린 문제가 없습니다.")

    return "\n".join(lines) + "\n"

File: web/src/grading.js
// Pure grading/validation logic, mirroring app/grading.py from the Python version.
// Kept dependency-free so it's easy to reason about and reuse from App.jsx.

export const DEFAULT_QUESTION_COUNT = 100
export const OPTIONS = ['A', 'B', 'C', 'D']

export function normalizeAnswerString(raw) {
  return raw.trim().toUpperCase().replace(/\s+/g, '')
}

// When `total` is given, raw must be exactly that many characters (used to
// check submitted answers against an already-registered key's length). When
// omitted, any non-empty string of valid option characters is accepted — the
// key's own length becomes the question count to grade.
export function validateAnswerString(raw, total = null) {
  if (!raw) {
    return '정답을 입력해주세요.'
  }
  if (total !== null && raw.length !== total) {
    return `정답은 정확히 ${total}자여야 합니다. (현재: ${raw.length}자)`
  }
  for (const char of raw) {
    if (!OPTIONS.includes(char)) {
      return `유효하지 않은 문자: '${char}'`
    }
  }
  return null
}

export function parseAnswerString(raw) {
  const map = {}
  for (let i = 0; i < raw.length; i++) {
    map[i + 1] = raw[i]
  }
  return map
}

export function grade(userAnswers, answerKey, total = null) {
  const effectiveTotal = total ?? Object.keys(answerKey).length
  let correctCount = 0
  const wrongQuestions = []
  for (let q = 1; q <= effectiveTotal; q++) {
    if (userAnswers[q] === answerKey[q]) {
      correctCount += 1
    } else {
      wrongQuestions.push(q)
    }
  }
  return { correctCount, wrongQuestions }
}

export function buildResultText({
  timestamp,
  correctCount,
  total,
  userAnswers,
  answerKey,
  wrongQuestions,
  notes,
}) {
  const accuracy = ((correctCount / total) * 100).toFixed(1)
  const lines = [
    '=========================================',
    '           토익 채점 및 오답 노트          ',
    '=========================================',
    `일시: ${timestamp}`,
    `점수: ${correctCount} / ${total} (${accuracy}%)`,
    '',
    '[ 전체 답안 현황 ]',
  ]

  for (let i = 1; i <= total; i++) {
    const isCorrect = userAnswers[i] === answerKey[i] ? 'O' : 'X'
    lines.push(
      `${String(i).padStart(3, '0')}번 | 제출: ${userAnswers[i]} | 정답: ${answerKey[i]} | [${isCorrect}]`,
    )
  }

  lines.push('')
  lines.push('[ 오답 노트 ]')
  if (wrongQuestions.length > 0) {
    for (const q of wrongQuestions) {
      const reason = (notes[q] || '').trim() || '메모 없음'
      lines.push(`- ${String(q).padStart(3, '0')}번 (제출: ${userAnswers[q]} / 정답: ${answerKey[q]})`)
      lines.push(`  이유: ${reason}`)
    }
  } else {
    lines.push('틀린 문제가 없습니다.')
  }

  return lines.join('\n') + '\n'
}
